### PR TITLE
Annotations, pylint, docs

### PR DESCRIPTION
**Pull request by johny-b**
This PR resolves following JIRA ticket

 
- [APPS-214](https://golemproject.atlassian.net/browse/APPS-214)

### Why

We consider showing this library to the outside world. This requires high code quality and docs.

### What

1. Linter setup
2. Annotations
3. Changes required by `pylint`
4. README

I think there are quite a lot of `pylint: disable`. Maybe we could get rid of some of them?

I moved `Cluster` to a separate file so the diff is not pretty. There were no changes in the logic, I only made some attributes private.

## Testing instructions

The same as in https://github.com/golemfactory/ya-httpx-client/pull/4 (just change the branch names).
I don't think testing is required for this PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,37 @@
...
 
-## Quickstart
+## Introduction
+
+The documentation assumes reader knows what [Golem](https://www.golem.network/) is and understand basics of the 
+development of [yapapi-based services](https://handbook.golem.network/requestor-tutorials/service-development).
+
+Features:
+
+1. Deploy a HTTP server on a Golem provider in an easy way. There are no requirements on the server implementation
```
Comment (Wiezzel) [inline README.md]:
```suggestion
1. Deploy an HTTP server on a Golem provider in an easy way. There are no requirements on the server implementation
```
It's pronounced *aitch-tee-tee-pee* so I always use 'an' before 'HTTP'. [Random guy on the internet agrees.](https://www.quora.com/A-http-or-an-http-which-is-the-correct-grammatical-usage)

File: README.md [hunk 2]
```diff
@@ -64,11 +95,22 @@ class LoadBalancer:
...
-If the last case, we have no control on how often `int(load_balancer_object)` will be called, so the implementation
+If the last case, we have no control on when and how often `int(load_balancer_object)` will be called, so the implementation
 should be a little more clever, at least to avoid too frequent changes - check [SimpleLoadBalancer](ya_httpx_client/provider_auto_balance.py)
 for an example.
 
 NOTE: setting size to anything other than an integer should be considered an experimental feature.
+
+## Future development
+
+1. Additional example: `requestor-proxy` - a HTTP server running on **requestor** that serves as a proxy to server(s) running on the provider(s).
```
Comment (Wiezzel) [inline README.md]:
```suggestion
1. Additional example: `requestor-proxy` - am HTTP server running on **requestor** that serves as a proxy to server(s) running on the provider(s).
```

File: README.md [hunk 3]
```diff
@@ -64,11 +95,22 @@ class LoadBalancer:
...
+2. Currently when the service stops it is restarted on another provider. This makes sense only for stateless services, but there is no way to turn this off.
```
Comment (Wiezzel) [inline README.md]:
```suggestion
2. Currently, when the service stops, it is restarted on another provider. This makes sense only for stateless services, but there is no way to turn this off.
```

File: README.md [hunk 4]
```diff
@@ -64,11 +95,22 @@ class LoadBalancer:
...
+   The developer should decide if they want to restart the service or not (or maybe stop the execution?).
+
+## Known issues
+
+1. Communication with providers is quite slow (1-2s for each request). This will be fixed when the new communication options are implemented in [yagna](https://github.com/golemfactory/yagna)
+   (hopefully in the next release).
```
Comment (Wiezzel) [inline README.md]:
```suggestion
   (hopefully in the next major release).
```

File: tests/test_serializable_request.py [hunk 5]
```diff
@@ -67,8 +67,8 @@ async def client():
     #   TODO command arg for subnet_tag
     session = Session(EXECUTOR_CFG)
     session.startup(**STARTUP_CFG)(echo_server_startup)
-    async with session.client() as client:
-        yield client
+    async with session.client() as cli:
```
Comment (Wiezzel) [inline tests/test_serializable_request.py:70]: `cli` is a popular acronym for 'command line interface' so it's not the best choice in my opinion. What was wrong with `client`?

File: ya_httpx_client/cluster.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+if TYPE_CHECKING:
+    from typing import Callable, Union, SupportsInt, List, Type, Optional
+    from yapapi_service_manager import ServiceManager
+    from yapapi import WorkContext
+
+
+class Cluster:
+    # pylint: disable=too-many-instance-attributes
+    '''
+    Q: Here is a Cluster, and we have a `yapapi.Cluster`. Both seem to wrap a bunch of services.
```
Comment (pnowosie) [inline ya_httpx_client/cluster.py]: I'm not sure what `Cluster` is referred here. 

```suggestion
    Q: Here is a Cluster class, and we have a `yapapi.Cluster`. Both seem to wrap a bunch of services.
```

File: ya_httpx_client/cluster.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+from .service_base import ServiceBase
+
+if TYPE_CHECKING:
+    from typing import Callable, Union, SupportsInt, List, Type, Optional
+    from yapapi_service_manager import ServiceManager
+    from yapapi import WorkContext
+
+
+class Cluster:
+    # pylint: disable=too-many-instance-attributes
```
Comment (Wiezzel) [inline ya_httpx_client/cluster.py]: Better increase the number of allowed attributes in `.pylintrc`:
```
[DESIGN]
max-attributes=12
```

File: ya_httpx_client/cluster.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    '''
+    Q: Here is a Cluster, and we have a `yapapi.Cluster`. Both seem to wrap a bunch of services.
+       Do we need two separate clusters for this?
+    A: In the future, this Cluster and `yapapi.Cluster` and also `yapapi_service_manager.ServiceWrapper`
+       should be merged into the same cluster-wrapper-thingy, but we can't do this without yapapi-side modifications.
+       Currently I don't think `yapapi.Cluster` could be used instead of this one.
```
Comment (pnowosie) [inline ya_httpx_client/cluster.py]:
```suggestion
       Currently I don't think the instance of `yapapi.Cluster` class could be used instead of this one.
```

File: ya_httpx_client/cluster.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            if service_wrapper.status in ('pending', 'starting'):
+                print(f"waiting for the service, current status: {service_wrapper.status}")
+            elif service_wrapper.status == 'running':
+                pass
+            else:
+                print(f"Replacing service on {service_wrapper.service.provider_name} - it is {service_wrapper.status}")
+                service_wrapper.service.restart_failed_request()
+
+                #   TODO: We don't stop the old service_wrapper, because it is dead either way.
+                #         This way we distinguish "failed" and "stopped" wrappers (although we don't
```
Comment (pnowosie) [inline ya_httpx_client/cluster.py]:
```suggestion
                #         We can distinguish "stopped" wrappers from  "failed"  (although we don't
```

File: ya_httpx_client/provider_auto_balance.py [hunk 10]
```diff
@@ -1,10 +1,10 @@
 from typing import TYPE_CHECKING
 from datetime import datetime
 if TYPE_CHECKING:
-    from ya_httpx_client.session import Cluster
+    from ya_httpx_client.cluster import Cluster
 
 
-class SimpleLoadBalancer:
+class SimpleLoadBalancer:  # pylint: disable=too-few-public-methods
```
Comment (Wiezzel) [inline ya_httpx_client/provider_auto_balance.py]: Again, a global disable is fine here so either add `too-few-public-methods` to the disable list in `.pylintrc` or put in this:
```
[DESIGN]
min-public-methods=0
```

**Review by pnowosie — APPROVED** [review]
Suggestions to wording

File: ya_httpx_client/service_base.py [hunk 11]
```diff
@@ -12,17 +12,17 @@ class ServiceBase(Service):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-        self.queue = self._yhc_cluster.request_queue
+        self.queue = self._yhc_cluster.request_queue  # pylint: disable=no-member
```
Comment (Wiezzel) [inline ya_httpx_client/service_base.py:15]: Why is this needed? Seems like something to be fixed rather than ignored.

**Comment by pnowosie** [comment]
Also consider to change PR title to something readable without branch naming conventions stuff

File: ya_httpx_client/session.py [hunk 12]
```diff
@@ -115,6 +21,7 @@ def __init__(self, request_queue):
         self.request_queue = request_queue
 
     async def handle_async_request(self, method, url, headers, stream, extensions):
+        # pylint: disable=too-many-arguments
```
Comment (Wiezzel) [inline ya_httpx_client/session.py:24]: In `.pylintrc`:
```
[DESIGN]
max-args=6
```

File: tests/test_serializable_request.py (hunk 5 above)
Comment (johny-b) [reply tests/test_serializable_request.py:70]: It's the same as the function name and `pylint` doesn't like this:
```
tests/test_serializable_request.py:70:35: W0621: Redefining name 'client' from outer scope (line 66) (redefined-outer-name)
```

Comment (Wiezzel) [reply tests/test_serializable_request.py:70]: Oh, okay then.

File: ya_httpx_client/cluster.py (hunk 8 above)
Comment (Wiezzel) [reply ya_httpx_client/cluster.py]: You probably meant 'an instance' here.

File: ya_httpx_client/session.py (hunk 12 above)
Comment (johny-b) [reply ya_httpx_client/session.py:24]: As discussed, I left it without changes

Comment (Wiezzel) [reply ya_httpx_client/session.py:24]: On second thoughts, better leave it as it is (a local exception to the rule).

File: ya_httpx_client/service_base.py (hunk 11 above)
Comment (johny-b) [reply ya_httpx_client/service_base.py:15]: This is a class created in a dynamic way (because the released `yapapi` is missing instance parametrization), so here we use in the general class an attribute that is defined only in the class that inherits from it.

I don't know how to fix this in a pretty way? Add a class attribute that will be always overwritten? That's not pretty at all.

I would leave this as-is because this will either way change with newer `yapapi`.

Comment (johny-b) [reply ya_httpx_client/service_base.py:15]: https://github.com/golemfactory/ya-httpx-client/issues/8
